apiserver/charms.go

```diff
@@ -416,55 +426,59 @@ func (h *charmsHandler) repackageAndUploadCharm(st *state.State, archive *charm.
 }
 
 // processGet handles a charm file GET request after authentication.
-// It returns the bundle path, the requested file path (if any) and an error.
-func (h *charmsHandler) processGet(r *http.Request, st *state.State) (string, string, error) {
+// It returns the bundle path, the requested file path (if any), whether the
+// default charm icon has been requested and an error.
+func (h *charmsHandler) processGet(r *http.Request, st *state.State) (string, string, bool, error) {
 	query := r.URL.Query()
 
 	// Retrieve and validate query parameters.
 	curlString := query.Get("url")
 	if curlString == "" {
-		return "", "", fmt.Errorf("expected url=CharmURL query argument")
+		return "", "", false, fmt.Errorf("expected url=CharmURL query argument")
```

Comment:
I saw a nice idiom recently that can reduce the boilerplate in this kind
of case. Define a local variable:

    errRet := func(err error) (string, string, bool, error) {
         return "", "", false, err
    }

then all your returns can just be:

    return errRet(fmt.Errorf(....))

and if you want to change the return parameters, you just need to change it one
place. Useful for large value-type names too.